diff -ur mate-panel-1.22.1/applets/wncklet/wnck-tsol.c mate-panel-1.22.1/applets/wncklet/wnck-tsol.c
--- mate-panel-1.22.1/applets/wncklet/wnck-tsol.c	2020-05-20 21:09:21.853321460 +0000
+++ mate-panel-1.22.1/applets/wncklet/wnck-tsol.c	2020-05-26 16:37:28.937449688 +0000
@@ -1,3 +1,7 @@
+/*
+ * Copyright (c) 2020, Dynamic Systems, Inc.
+ */
+
 #include  <config.h>
 #ifdef HAVE_LIBGNOMETSOL
 
diff -ur mate-panel-1.22.1/applets/wncklet/wnck-tsol.h mate-panel-1.22.1/applets/wncklet/wnck-tsol.h
--- mate-panel-1.22.1/applets/wncklet/wnck-tsol.h	2020-05-20 21:09:21.854007098 +0000
+++ mate-panel-1.22.1/applets/wncklet/wnck-tsol.h	2020-05-26 16:37:57.014252513 +0000
@@ -1,3 +1,7 @@
+/*
+ * Copyright (c) 2020, Dynamic Systems, Inc.
+ */
+
 #ifndef __WNCK_TSOL_H__
 #define __WNCK_TSOL_H__
 
diff -ur mate-panel-1.22.1/applets/wncklet/workspace-switcher.c mate-panel-1.22.1/applets/wncklet/workspace-switcher.c
--- mate-panel-1.22.1/applets/wncklet/workspace-switcher.c	2020-05-20 21:09:21.855691800 +0000
+++ mate-panel-1.22.1/applets/wncklet/workspace-switcher.c	2020-05-26 16:38:23.748836780 +0000
@@ -3,6 +3,7 @@
  * libwnck based pager applet.
  * (C) 2001 Alexander Larsson
  * (C) 2001 Red Hat, Inc
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Authors: Alexander Larsson
  *
diff -ur mate-panel-1.22.1/mate-panel/launcher.c mate-panel-1.22.1/mate-panel/launcher.c
--- mate-panel-1.22.1/mate-panel/launcher.c	2020-05-20 21:09:21.857224258 +0000
+++ mate-panel-1.22.1/mate-panel/launcher.c	2020-05-26 16:38:39.716654661 +0000
@@ -2,6 +2,7 @@
  * MATE panel launcher module.
  * (C) 1997,1998,1999,2000 The Free Software Foundation
  * (C) 2000 Eazel, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Authors: Miguel de Icaza
  *          Federico Mena
diff -ur mate-panel-1.22.1/mate-panel/libmate-panel-applet-private/panel-applets-manager-dbus.c mate-panel-1.22.1/mate-panel/libmate-panel-applet-private/panel-applets-manager-dbus.c
--- mate-panel-1.22.1/mate-panel/libmate-panel-applet-private/panel-applets-manager-dbus.c	2020-05-20 21:09:21.857744013 +0000
+++ mate-panel-1.22.1/mate-panel/libmate-panel-applet-private/panel-applets-manager-dbus.c	2020-05-26 16:38:46.341972692 +0000
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2010 Carlos Garcia Campos <carlosgc@gnome.org>
  * Copyright (C) 2010 Vincent Untz <vuntz@gnome.org>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/main.c mate-panel-1.22.1/mate-panel/main.c
--- mate-panel-1.22.1/mate-panel/main.c	2020-05-20 21:09:21.858128727 +0000
+++ mate-panel-1.22.1/mate-panel/main.c	2020-05-26 16:38:53.969066881 +0000
@@ -1,6 +1,7 @@
 /* Mate panel: Initialization routines
  * (C) 1997,1998,1999,2000 the Free Software Foundation
  * (C) 2000 Eazel, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Authors: Federico Mena
  *          Miguel de Icaza
diff -ur mate-panel-1.22.1/mate-panel/panel-action-button.c mate-panel-1.22.1/mate-panel/panel-action-button.c
--- mate-panel-1.22.1/mate-panel/panel-action-button.c	2020-05-20 21:09:21.858641212 +0000
+++ mate-panel-1.22.1/mate-panel/panel-action-button.c	2020-05-26 16:39:02.542919832 +0000
@@ -4,6 +4,7 @@
  *
  * Copyright (C) 2002 Sun Microsystems, Inc.
  * Copyright (C) 2004 Red Hat, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-addto.c mate-panel-1.22.1/mate-panel/panel-addto.c
--- mate-panel-1.22.1/mate-panel/panel-addto.c	2020-05-20 21:09:21.859232255 +0000
+++ mate-panel-1.22.1/mate-panel/panel-addto.c	2020-05-26 16:39:10.908908528 +0000
@@ -2,6 +2,7 @@
  * panel-addto.c:
  *
  * Copyright (C) 2004 Vincent Untz
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-applet-frame.c mate-panel-1.22.1/mate-panel/panel-applet-frame.c
--- mate-panel-1.22.1/mate-panel/panel-applet-frame.c	2020-05-20 21:09:21.859805126 +0000
+++ mate-panel-1.22.1/mate-panel/panel-applet-frame.c	2020-05-26 16:39:21.103417229 +0000
@@ -4,6 +4,7 @@
  *
  * Copyright (C) 2010 Carlos Garcia Campos <carlosgc@gnome.org>
  * Copyright (C) 2001 - 2003 Sun Microsystems, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-applet-info.c mate-panel-1.22.1/mate-panel/panel-applet-info.c
--- mate-panel-1.22.1/mate-panel/panel-applet-info.c	2020-05-20 21:09:21.860339037 +0000
+++ mate-panel-1.22.1/mate-panel/panel-applet-info.c	2020-05-26 16:39:28.230975932 +0000
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2010 Carlos Garcia Campos <carlosgc@gnome.org>
  * Copyright (C) 2010 Vincent Untz <vuntz@gnome.org>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-applet-info.h mate-panel-1.22.1/mate-panel/panel-applet-info.h
--- mate-panel-1.22.1/mate-panel/panel-applet-info.h	2020-05-20 21:09:21.860735894 +0000
+++ mate-panel-1.22.1/mate-panel/panel-applet-info.h	2020-05-26 16:39:34.725138061 +0000
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2010 Carlos Garcia Campos <carlosgc@gnome.org>
  * Copyright (C) 2010 Vincent Untz <vuntz@gnome.org>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-lockdown.c mate-panel-1.22.1/mate-panel/panel-lockdown.c
--- mate-panel-1.22.1/mate-panel/panel-lockdown.c	2020-05-20 21:09:21.861389666 +0000
+++ mate-panel-1.22.1/mate-panel/panel-lockdown.c	2020-05-26 16:39:42.068176582 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*- */
 /*
  * Copyright (C) 2004 Sun Microsystems, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-lockdown.h mate-panel-1.22.1/mate-panel/panel-lockdown.h
--- mate-panel-1.22.1/mate-panel/panel-lockdown.h	2020-05-20 21:09:21.861838926 +0000
+++ mate-panel-1.22.1/mate-panel/panel-lockdown.h	2020-05-26 16:39:47.761335588 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*- */
 /*
  * Copyright (C) 2004 Sun Microsystems, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-menu-items.c mate-panel-1.22.1/mate-panel/panel-menu-items.c
--- mate-panel-1.22.1/mate-panel/panel-menu-items.c	2020-05-20 21:09:21.862712431 +0000
+++ mate-panel-1.22.1/mate-panel/panel-menu-items.c	2020-05-26 16:39:54.291948175 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2005 Vincent Untz
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-multiscreen.c mate-panel-1.22.1/mate-panel/panel-multiscreen.c
--- mate-panel-1.22.1/mate-panel/panel-multiscreen.c	2020-05-20 21:09:21.886645308 +0000
+++ mate-panel-1.22.1/mate-panel/panel-multiscreen.c	2020-05-26 16:40:34.824969777 +0000
@@ -3,7 +3,7 @@
  *
  * Copyright (C) 2001 George Lebl <jirka@5z.com>
  *               2002 Sun Microsystems Inc.
- *
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
  * published by the Free Software Foundation; either version 2 of the
diff -ur mate-panel-1.22.1/mate-panel/panel-run-dialog.c mate-panel-1.22.1/mate-panel/panel-run-dialog.c
--- mate-panel-1.22.1/mate-panel/panel-run-dialog.c	2020-05-20 21:09:21.863964076 +0000
+++ mate-panel-1.22.1/mate-panel/panel-run-dialog.c	2020-05-26 16:40:50.688178687 +0000
@@ -2,6 +2,7 @@
  * panel-run-dialog.c:
  *
  * Copyright (C) 2003 Frank Worsley <fworsley@shaw.ca>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur mate-panel-1.22.1/mate-panel/panel-solaris.c mate-panel-1.22.1/mate-panel/panel-solaris.c
--- mate-panel-1.22.1/mate-panel/panel-solaris.c	2020-05-20 21:09:21.864493041 +0000
+++ mate-panel-1.22.1/mate-panel/panel-solaris.c	2020-05-26 16:41:31.821173264 +0000
@@ -1,4 +1,7 @@
 /*
+ * Copyright (c) 2020, Dynamic Systems, Inc.
+ */
+/*
  * Note that the following code is in three patches:
  * - gnome-panel-XX-rbac.diff (filter_with_rbac)
  * - gnome-menus-XX-rbac.diff (filter_with_rbac)
diff -ur mate-panel-1.22.1/mate-panel/panel-solaris.h mate-panel-1.22.1/mate-panel/panel-solaris.h
--- mate-panel-1.22.1/mate-panel/panel-solaris.h	2020-05-20 21:09:21.865190652 +0000
+++ mate-panel-1.22.1/mate-panel/panel-solaris.h	2020-05-26 16:41:53.510531804 +0000
@@ -1,3 +1,6 @@
+/*
+ * Copyright (c) 2020, Dynamic Systems, Inc.
+ */
 
 #ifndef __MENU_SOLARIS_H__
 #define __MENU_SOLARIS_H__
diff -ur mate-panel-1.22.1/mate-panel/panel-struts.c mate-panel-1.22.1/mate-panel/panel-struts.c
--- mate-panel-1.22.1/mate-panel/panel-struts.c	2020-05-20 21:09:21.866013287 +0000
+++ mate-panel-1.22.1/mate-panel/panel-struts.c	2020-05-26 16:42:01.992069947 +0000
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2003 Sun Microsystems, Inc.
  * Copyright (C) 2003,2004 Rob Adams
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
