diff -ur nimbus-172733afb6156c59fcd57484854090b798460fd9/icons/96x96/places/Makefile.am nimbus/icons/96x96/places/Makefile.am
--- nimbus-172733afb6156c59fcd57484854090b798460fd9/icons/96x96/places/Makefile.am	2020-06-08 15:40:01.396726576 +0000
+++ nimbus-172733afb6156c59fcd57484854090b798460fd9/icons/96x96/places/Makefile.am	2020-06-08 15:40:01.396726576 +0000
@@ -16,6 +16,7 @@
 		gnome-fs-ssh.png 		\
 		network-server.png 		\
 		network-workgroup.png 		\
+		start-here.png 		\
 		user-desktop.png 		\
 		user-home.png 		\
 		user-trash-full.png 		\
