diff -ur tigervnc-1.7.1/unix/xserver/mi/miinitext.c tigervnc-1.7.1/unix/xserver/mi/miinitext.c
--- tigervnc-1.7.1/unix/xserver/mi/miinitext.c	2020-05-01 17:30:16.741411621 +0000
+++ tigervnc-1.7.1/unix/xserver/mi/miinitext.c	2020-05-26 17:03:54.858431483 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
