--- a/driver/Makefile.in	2020-05-19 15:23:46.620457227 +0000
+++ b/driver/Makefile.in	2020-05-19 15:24:39.559536972 +0000
@@ -151,7 +151,7 @@
 
 ICON_SRC        = $(UTILS_SRC)/images
 LOGO		= $(ICON_SRC)/logo-50.xpm
-GTK_ICONS	= $(ICON_SRC)/screensaver-*.png
+GTK_ICONS	= $(ICON_SRC)/screensaver-*.png $(ICON_SRC)/logo*.gif
 
 DEMO_UTIL_SRCS	= $(UTILS_SRC)/resources.c $(UTILS_SRC)/usleep.c \
 		  $(UTILS_SRC)/visual.c $(XMU_SRCS)
