diff -ur xscreensaver-5.39/driver/Makefile.in xscreensaver-5.39/driver/Makefile.in
--- xscreensaver-5.39/driver/Makefile.in	2020-05-19 15:28:32.758299837 +0000
+++ xscreensaver-5.39/driver/Makefile.in	2020-05-26 16:45:15.549942371 +0000
@@ -1,4 +1,5 @@
 # driver/Makefile.in --- xscreensaver, Copyright (c) 1997-2010 Jamie Zawinski.
+# Copyright (c) 2020, Dynamic Systems, Inc.
 # the `../configure' script generates `driver/Makefile' from this file.
 
 @SET_MAKE@
