diff -ur xorg-server-1.20.3/Xext/security.c xorg-server-1.20.3/Xext/security.c
--- xorg-server-1.20.3/Xext/security.c	2020-05-26 15:03:02.098103971 +0000
+++ xorg-server-1.20.3/Xext/security.c	2020-05-26 16:54:35.633984509 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/Xext/securitysrv.h xorg-server-1.20.3/Xext/securitysrv.h
--- xorg-server-1.20.3/Xext/securitysrv.h	2020-05-26 15:03:02.098824086 +0000
+++ xorg-server-1.20.3/Xext/securitysrv.h	2020-05-26 16:54:40.354862274 +0000
@@ -1,5 +1,6 @@
 /*
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/configure.ac xorg-server-1.20.3/configure.ac
--- xorg-server-1.20.3/configure.ac	2020-05-26 15:03:02.100176580 +0000
+++ xorg-server-1.20.3/configure.ac	2020-05-26 16:55:01.022837793 +0000
@@ -1,4 +1,5 @@
 dnl Copyright © 2003-2007 Keith Packard, Daniel Stone
+dnl Copyright (c) 2020, Dynamic Systems, Inc.
 dnl
 dnl Permission is hereby granted, free of charge, to any person obtaining a
 dnl copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.3/dix/dispatch.c xorg-server-1.20.3/dix/dispatch.c
--- xorg-server-1.20.3/dix/dispatch.c	2020-05-26 15:03:02.101639573 +0000
+++ xorg-server-1.20.3/dix/dispatch.c	2020-05-26 16:55:24.832781884 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/dix/events.c xorg-server-1.20.3/dix/events.c
--- xorg-server-1.20.3/dix/events.c	2020-05-26 15:03:02.103242280 +0000
+++ xorg-server-1.20.3/dix/events.c	2020-05-26 16:55:30.355783726 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/dix/window.c xorg-server-1.20.3/dix/window.c
--- xorg-server-1.20.3/dix/window.c	2020-05-26 15:03:02.104551154 +0000
+++ xorg-server-1.20.3/dix/window.c	2020-05-26 16:55:35.684203204 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright (c) 2006, Red Hat, Inc.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission is hereby granted, free of charge, to any person obtaining a
 copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.3/hw/xfree86/common/xf86Config.c xorg-server-1.20.3/hw/xfree86/common/xf86Config.c
--- xorg-server-1.20.3/hw/xfree86/common/xf86Config.c	2020-05-26 15:03:02.105613454 +0000
+++ xorg-server-1.20.3/hw/xfree86/common/xf86Config.c	2020-05-26 16:56:02.930244692 +0000
@@ -7,6 +7,7 @@
 /*
  * Copyright 1992-2003 by The XFree86 Project, Inc.
  * Copyright 1997 by Metro Link, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.3/include/dix.h xorg-server-1.20.3/include/dix.h
--- xorg-server-1.20.3/include/dix.h	2020-05-26 15:03:02.107167371 +0000
+++ xorg-server-1.20.3/include/dix.h	2020-05-26 16:56:19.592699931 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/include/dixstruct.h xorg-server-1.20.3/include/dixstruct.h
--- xorg-server-1.20.3/include/dixstruct.h	2020-05-26 15:03:02.107987668 +0000
+++ xorg-server-1.20.3/include/dixstruct.h	2020-05-26 16:56:32.358877128 +0000
@@ -1,5 +1,6 @@
 /***********************************************************
 Copyright 1987 by Digital Equipment Corporation, Maynard, Massachusetts.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.20.3/include/extnsionst.h xorg-server-1.20.3/include/extnsionst.h
--- xorg-server-1.20.3/include/extnsionst.h	2020-05-26 15:03:02.110417907 +0000
+++ xorg-server-1.20.3/include/extnsionst.h	2020-05-26 16:56:42.805570185 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/include/input.h xorg-server-1.20.3/include/input.h
--- xorg-server-1.20.3/include/input.h	2020-05-26 15:03:02.108804138 +0000
+++ xorg-server-1.20.3/include/input.h	2020-05-26 16:56:48.374405233 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/include/swaprep.h xorg-server-1.20.3/include/swaprep.h
--- xorg-server-1.20.3/include/swaprep.h	2020-05-26 15:03:02.109579066 +0000
+++ xorg-server-1.20.3/include/swaprep.h	2020-05-26 16:56:54.700135213 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1996 by Thomas E. Dickey <dickey@clark.net>
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.20.3/include/window.h xorg-server-1.20.3/include/window.h
--- xorg-server-1.20.3/include/window.h	2020-05-26 15:03:02.111188880 +0000
+++ xorg-server-1.20.3/include/window.h	2020-05-26 16:57:01.332764380 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/mi/miinitext.c xorg-server-1.20.3/mi/miinitext.c
--- xorg-server-1.20.3/mi/miinitext.c	2020-05-26 15:03:02.111985848 +0000
+++ xorg-server-1.20.3/mi/miinitext.c	2020-05-26 16:57:06.328640460 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/os/connection.c xorg-server-1.20.3/os/connection.c
--- xorg-server-1.20.3/os/connection.c	2020-05-26 15:03:02.112844413 +0000
+++ xorg-server-1.20.3/os/connection.c	2020-05-26 16:57:11.372416513 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.3/randr/randrstr.h xorg-server-1.20.3/randr/randrstr.h
--- xorg-server-1.20.3/randr/randrstr.h	2020-05-26 15:03:02.113702178 +0000
+++ xorg-server-1.20.3/randr/randrstr.h	2020-05-26 16:57:17.923933168 +0000
@@ -3,6 +3,7 @@
  * Copyright © 2002 Hewlett-Packard Company
  * Copyright © 2006 Intel Corporation
  * Copyright © 2008 Red Hat, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission to use, copy, modify, distribute, and sell this software and its
  * documentation for any purpose is hereby granted without fee, provided that
