diff -ur xorg-server-1.20.6/Xext/security.c xorg-server-1.20.6/Xext/security.c
--- xorg-server-1.20.6/Xext/security.c	2020-05-26 15:02:17.058312457 +0000
+++ xorg-server-1.20.6/Xext/security.c	2020-05-26 16:57:27.084059545 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/Xext/securitysrv.h xorg-server-1.20.6/Xext/securitysrv.h
--- xorg-server-1.20.6/Xext/securitysrv.h	2020-05-26 15:02:17.059052202 +0000
+++ xorg-server-1.20.6/Xext/securitysrv.h	2020-05-26 16:57:32.619177070 +0000
@@ -1,5 +1,6 @@
 /*
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/configure.ac xorg-server-1.20.6/configure.ac
--- xorg-server-1.20.6/configure.ac	2020-05-26 15:02:17.060717370 +0000
+++ xorg-server-1.20.6/configure.ac	2020-05-26 16:57:42.093558777 +0000
@@ -1,4 +1,5 @@
 dnl Copyright © 2003-2007 Keith Packard, Daniel Stone
+dnl Copyright (c) 2020, Dynamic Systems, Inc.
 dnl
 dnl Permission is hereby granted, free of charge, to any person obtaining a
 dnl copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.6/dix/dispatch.c xorg-server-1.20.6/dix/dispatch.c
--- xorg-server-1.20.6/dix/dispatch.c	2020-05-26 15:02:17.062159723 +0000
+++ xorg-server-1.20.6/dix/dispatch.c	2020-05-26 16:57:48.589165229 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/dix/events.c xorg-server-1.20.6/dix/events.c
--- xorg-server-1.20.6/dix/events.c	2020-05-26 15:02:17.063790699 +0000
+++ xorg-server-1.20.6/dix/events.c	2020-05-26 16:57:53.386458407 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/dix/window.c xorg-server-1.20.6/dix/window.c
--- xorg-server-1.20.6/dix/window.c	2020-05-26 15:02:17.065077106 +0000
+++ xorg-server-1.20.6/dix/window.c	2020-05-26 16:57:58.289160023 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright (c) 2006, Red Hat, Inc.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission is hereby granted, free of charge, to any person obtaining a
 copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.6/hw/xfree86/common/xf86Config.c xorg-server-1.20.6/hw/xfree86/common/xf86Config.c
--- xorg-server-1.20.6/hw/xfree86/common/xf86Config.c	2020-05-26 15:02:17.066179219 +0000
+++ xorg-server-1.20.6/hw/xfree86/common/xf86Config.c	2020-05-26 16:58:05.500671400 +0000
@@ -7,6 +7,7 @@
 /*
  * Copyright 1992-2003 by The XFree86 Project, Inc.
  * Copyright 1997 by Metro Link, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.20.6/include/dix.h xorg-server-1.20.6/include/dix.h
--- xorg-server-1.20.6/include/dix.h	2020-05-26 15:02:17.067762361 +0000
+++ xorg-server-1.20.6/include/dix.h	2020-05-26 16:58:14.620725583 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/include/dixstruct.h xorg-server-1.20.6/include/dixstruct.h
--- xorg-server-1.20.6/include/dixstruct.h	2020-05-26 15:02:17.068567815 +0000
+++ xorg-server-1.20.6/include/dixstruct.h	2020-05-26 16:58:20.428801307 +0000
@@ -1,5 +1,6 @@
 /***********************************************************
 Copyright 1987 by Digital Equipment Corporation, Maynard, Massachusetts.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.20.6/include/extnsionst.h xorg-server-1.20.6/include/extnsionst.h
--- xorg-server-1.20.6/include/extnsionst.h	2020-05-26 15:02:17.071153260 +0000
+++ xorg-server-1.20.6/include/extnsionst.h	2020-05-26 16:58:25.442417013 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/include/input.h xorg-server-1.20.6/include/input.h
--- xorg-server-1.20.6/include/input.h	2020-05-26 15:02:17.069422258 +0000
+++ xorg-server-1.20.6/include/input.h	2020-05-26 16:58:30.995922184 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/include/swaprep.h xorg-server-1.20.6/include/swaprep.h
--- xorg-server-1.20.6/include/swaprep.h	2020-05-26 15:02:17.070259858 +0000
+++ xorg-server-1.20.6/include/swaprep.h	2020-05-26 16:58:52.979596045 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1996 by Thomas E. Dickey <dickey@clark.net>
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.20.6/include/window.h xorg-server-1.20.6/include/window.h
--- xorg-server-1.20.6/include/window.h	2020-05-26 15:02:17.071936892 +0000
+++ xorg-server-1.20.6/include/window.h	2020-05-26 16:58:57.624423822 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/mi/miinitext.c xorg-server-1.20.6/mi/miinitext.c
--- xorg-server-1.20.6/mi/miinitext.c	2020-05-26 15:02:17.072778692 +0000
+++ xorg-server-1.20.6/mi/miinitext.c	2020-05-26 16:59:03.002921288 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/os/connection.c xorg-server-1.20.6/os/connection.c
--- xorg-server-1.20.6/os/connection.c	2020-05-26 15:02:17.073677036 +0000
+++ xorg-server-1.20.6/os/connection.c	2020-05-26 16:59:07.783665343 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.20.6/randr/randrstr.h xorg-server-1.20.6/randr/randrstr.h
--- xorg-server-1.20.6/randr/randrstr.h	2020-05-26 15:02:17.074633436 +0000
+++ xorg-server-1.20.6/randr/randrstr.h	2020-05-26 16:59:14.094133367 +0000
@@ -3,6 +3,7 @@
  * Copyright © 2002 Hewlett-Packard Company
  * Copyright © 2006 Intel Corporation
  * Copyright © 2008 Red Hat, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission to use, copy, modify, distribute, and sell this software and its
  * documentation for any purpose is hereby granted without fee, provided that
