diff -urN data/sounds/Makefile.am data2/sounds/Makefile.am
--- a/data/sounds/Makefile.am	2020-08-06 00:09:45.574168969 +0000
+++ b/data/sounds/Makefile.am	2020-08-06 00:09:29.344314667 +0000
@@ -7,6 +7,7 @@
 	drip.ogg		\
 	glass.ogg		\
 	sonar.ogg		\
+	index.theme		\
 	$(NULL)
 
 metadata_in_files = mate-sounds-default.xml.in
diff -urN data/sounds/index.theme data2/sounds/index.theme
new file mode 100644
--- a/data/sounds/index.theme	1970-01-01 00:00:00.000000000 +0000
+++ b/data/sounds/index.theme	2020-08-06 00:09:29.344651247 +0000
@@ -0,0 +1,6 @@
+[Sound Theme]
+Name=Mate
+Directories=default
+
+[default]
+OutputProfile=default
