diff -ur xorg-server-1.19.5/Xext/security.c xorg-server-1.19.5/Xext/security.c
--- xorg-server-1.19.5/Xext/security.c	2020-05-01 17:06:18.218267916 +0000
+++ xorg-server-1.19.5/Xext/security.c	2020-05-26 17:01:01.523973093 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/Xext/securitysrv.h xorg-server-1.19.5/Xext/securitysrv.h
--- xorg-server-1.19.5/Xext/securitysrv.h	2020-05-01 17:06:18.231971440 +0000
+++ xorg-server-1.19.5/Xext/securitysrv.h	2020-05-26 17:01:15.276901987 +0000
@@ -1,5 +1,6 @@
 /*
 Copyright 1996, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/configure.ac xorg-server-1.19.5/configure.ac
--- xorg-server-1.19.5/configure.ac	2020-05-01 17:06:18.219921262 +0000
+++ xorg-server-1.19.5/configure.ac	2020-05-26 17:01:26.093874330 +0000
@@ -1,4 +1,5 @@
 dnl Copyright © 2003-2007 Keith Packard, Daniel Stone
+dnl Copyright (c) 2020, Dynamic Systems, Inc.
 dnl
 dnl Permission is hereby granted, free of charge, to any person obtaining a
 dnl copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/dix/dispatch.c xorg-server-1.19.5/dix/dispatch.c
--- xorg-server-1.19.5/dix/dispatch.c	2020-05-01 17:06:18.221299830 +0000
+++ xorg-server-1.19.5/dix/dispatch.c	2020-05-26 17:01:31.187913564 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/dix/events.c xorg-server-1.19.5/dix/events.c
--- xorg-server-1.19.5/dix/events.c	2020-05-01 17:06:18.222730463 +0000
+++ xorg-server-1.19.5/dix/events.c	2020-05-26 17:01:38.538607446 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/dix/window.c xorg-server-1.19.5/dix/window.c
--- xorg-server-1.19.5/dix/window.c	2020-05-01 17:06:18.224011565 +0000
+++ xorg-server-1.19.5/dix/window.c	2020-05-26 17:01:44.692981995 +0000
@@ -1,6 +1,7 @@
 /*
 
 Copyright (c) 2006, Red Hat, Inc.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission is hereby granted, free of charge, to any person obtaining a
 copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/hw/xfree86/common/xf86Config.c xorg-server-1.19.5/hw/xfree86/common/xf86Config.c
--- xorg-server-1.19.5/hw/xfree86/common/xf86Config.c	2020-05-01 17:06:18.233429630 +0000
+++ xorg-server-1.19.5/hw/xfree86/common/xf86Config.c	2020-05-26 17:01:52.255863313 +0000
@@ -7,6 +7,7 @@
 /*
  * Copyright 1992-2003 by The XFree86 Project, Inc.
  * Copyright 1997 by Metro Link, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/include/dix.h xorg-server-1.19.5/include/dix.h
--- xorg-server-1.19.5/include/dix.h	2020-05-01 17:06:18.225363767 +0000
+++ xorg-server-1.19.5/include/dix.h	2020-05-26 17:01:59.059729228 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/include/dixstruct.h xorg-server-1.19.5/include/dixstruct.h
--- xorg-server-1.19.5/include/dixstruct.h	2020-05-01 17:06:18.226057972 +0000
+++ xorg-server-1.19.5/include/dixstruct.h	2020-05-26 17:02:04.565431333 +0000
@@ -1,5 +1,6 @@
 /***********************************************************
 Copyright 1987 by Digital Equipment Corporation, Maynard, Massachusetts.
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.19.5/include/extnsionst.h xorg-server-1.19.5/include/extnsionst.h
--- xorg-server-1.19.5/include/extnsionst.h	2020-05-01 17:06:18.227484355 +0000
+++ xorg-server-1.19.5/include/extnsionst.h	2020-05-26 17:02:09.373169271 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/include/input.h xorg-server-1.19.5/include/input.h
--- xorg-server-1.19.5/include/input.h	2020-05-01 17:06:18.228170584 +0000
+++ xorg-server-1.19.5/include/input.h	2020-05-26 17:02:16.549529283 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/include/swaprep.h xorg-server-1.19.5/include/swaprep.h
--- xorg-server-1.19.5/include/swaprep.h	2020-05-01 17:06:18.226755287 +0000
+++ xorg-server-1.19.5/include/swaprep.h	2020-05-26 17:02:27.221851193 +0000
@@ -1,6 +1,7 @@
 /************************************************************
 
 Copyright 1996 by Thomas E. Dickey <dickey@clark.net>
+Copyright (c) 2020, Dynamic Systems, Inc.
 
                         All Rights Reserved
 
diff -ur xorg-server-1.19.5/include/window.h xorg-server-1.19.5/include/window.h
--- xorg-server-1.19.5/include/window.h	2020-05-01 17:06:18.228892064 +0000
+++ xorg-server-1.19.5/include/window.h	2020-05-26 17:02:21.307090436 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/mi/miinitext.c xorg-server-1.19.5/mi/miinitext.c
--- xorg-server-1.19.5/mi/miinitext.c	2020-05-01 17:06:18.232621882 +0000
+++ xorg-server-1.19.5/mi/miinitext.c	2020-05-26 17:02:31.636864993 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/os/connection.c xorg-server-1.19.5/os/connection.c
--- xorg-server-1.19.5/os/connection.c	2020-05-01 17:06:18.231376232 +0000
+++ xorg-server-1.19.5/os/connection.c	2020-05-26 17:02:37.651997400 +0000
@@ -1,6 +1,7 @@
 /***********************************************************
 
 Copyright 1987, 1989, 1998  The Open Group
+Copyright (c) 2020, Dynamic Systems, Inc.
 
 Permission to use, copy, modify, distribute, and sell this software and its
 documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/randr/randrstr.h xorg-server-1.19.5/randr/randrstr.h
--- xorg-server-1.19.5/randr/randrstr.h	2020-05-01 17:06:18.229784158 +0000
+++ xorg-server-1.19.5/randr/randrstr.h	2020-05-26 17:02:44.925019271 +0000
@@ -3,6 +3,7 @@
  * Copyright © 2002 Hewlett-Packard Company
  * Copyright © 2006 Intel Corporation
  * Copyright © 2008 Red Hat, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission to use, copy, modify, distribute, and sell this software and its
  * documentation for any purpose is hereby granted without fee, provided that
diff -ur xorg-server-1.19.5/tsol/TrustedExtensionsPolicy xorg-server-1.19.5/tsol/TrustedExtensionsPolicy
--- xorg-server-1.19.5/tsol/TrustedExtensionsPolicy	2020-05-01 17:06:18.249538239 +0000
+++ xorg-server-1.19.5/tsol/TrustedExtensionsPolicy	2020-05-26 17:02:50.726071884 +0000
@@ -1,5 +1,6 @@
 #
 # Copyright (c) 2004, 2007, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2020, Dynamic Systems, Inc.
 #
 # Permission is hereby granted, free of charge, to any person obtaining a
 # copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/auditwrite.c xorg-server-1.19.5/tsol/auditwrite.c
--- xorg-server-1.19.5/tsol/auditwrite.c	2020-05-01 17:06:18.251978740 +0000
+++ xorg-server-1.19.5/tsol/auditwrite.c	2020-05-26 17:02:56.140711310 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsol.h xorg-server-1.19.5/tsol/tsol.h
--- xorg-server-1.19.5/tsol/tsol.h	2020-05-01 17:06:18.252652168 +0000
+++ xorg-server-1.19.5/tsol/tsol.h	2020-05-26 17:03:02.736490339 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolextension.c xorg-server-1.19.5/tsol/tsolextension.c
--- xorg-server-1.19.5/tsol/tsolextension.c	2020-05-01 17:06:18.254374415 +0000
+++ xorg-server-1.19.5/tsol/tsolextension.c	2020-05-26 17:03:07.596030311 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolextension.h xorg-server-1.19.5/tsol/tsolextension.h
--- xorg-server-1.19.5/tsol/tsolextension.h	2020-05-01 17:06:18.255007174 +0000
+++ xorg-server-1.19.5/tsol/tsolextension.h	2020-05-26 17:03:12.147071255 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2010, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolinfo.h xorg-server-1.19.5/tsol/tsolinfo.h
--- xorg-server-1.19.5/tsol/tsolinfo.h	2020-05-01 17:06:18.255616172 +0000
+++ xorg-server-1.19.5/tsol/tsolinfo.h	2020-05-26 17:03:17.281633382 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2014, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolpolicy.c xorg-server-1.19.5/tsol/tsolpolicy.c
--- xorg-server-1.19.5/tsol/tsolpolicy.c	2020-05-01 17:06:18.257026175 +0000
+++ xorg-server-1.19.5/tsol/tsolpolicy.c	2020-05-26 17:03:22.723399976 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolprotocol.c xorg-server-1.19.5/tsol/tsolprotocol.c
--- xorg-server-1.19.5/tsol/tsolprotocol.c	2020-05-01 17:06:18.258169254 +0000
+++ xorg-server-1.19.5/tsol/tsolprotocol.c	2020-05-26 17:03:27.669506125 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
diff -ur xorg-server-1.19.5/tsol/tsolutils.c xorg-server-1.19.5/tsol/tsolutils.c
--- xorg-server-1.19.5/tsol/tsolutils.c	2020-05-01 17:06:18.258948569 +0000
+++ xorg-server-1.19.5/tsol/tsolutils.c	2020-05-26 17:03:33.977482336 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2004, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
