diff -ur caja-1.22.1/libcaja-private/caja-clipboard-monitor.c caja-1.22.1/libcaja-private/caja-clipboard-monitor.c
--- caja-1.22.1/libcaja-private/caja-clipboard-monitor.c	2020-04-24 23:05:52.998456819 +0000
+++ caja-1.22.1/libcaja-private/caja-clipboard-monitor.c	2020-05-26 16:29:47.462560987 +0000
@@ -3,6 +3,7 @@
    caja-clipboard-monitor.c: catch clipboard changes.
 
    Copyright (C) 2004 Red Hat, Inc.
+   Copyright (c) 2020, Dynamic Systems, Inc.
 
    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
diff -ur caja-1.22.1/libcaja-private/caja-file-utilities.c caja-1.22.1/libcaja-private/caja-file-utilities.c
--- caja-1.22.1/libcaja-private/caja-file-utilities.c	2020-04-24 23:05:52.999503645 +0000
+++ caja-1.22.1/libcaja-private/caja-file-utilities.c	2020-05-26 16:30:02.236767433 +0000
@@ -3,6 +3,7 @@
 /* caja-file-utilities.c - implementation of file manipulation routines.
 
    Copyright (C) 1999, 2000, 2001 Eazel, Inc.
+   Copyright (c) 2020, Dynamic Systems, Inc.
 
    The Mate Library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public License as
diff -ur caja-1.22.1/libcaja-private/caja-module.c caja-1.22.1/libcaja-private/caja-module.c
--- caja-1.22.1/libcaja-private/caja-module.c	2020-04-24 23:05:53.000343769 +0000
+++ caja-1.22.1/libcaja-private/caja-module.c	2020-05-26 16:30:13.705703546 +0000
@@ -2,6 +2,7 @@
  *  caja-module.h - Interface to caja extensions
  *
  *  Copyright (C) 2003 Novell, Inc.
+ *  Copyright (c) 2020, Dynamic Systems, Inc.
  *
  *  This library is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU Library General Public
diff -ur caja-1.22.1/src/caja-application.c caja-1.22.1/src/caja-application.c
--- caja-1.22.1/src/caja-application.c	2020-04-24 23:05:53.001507066 +0000
+++ caja-1.22.1/src/caja-application.c	2020-05-26 16:30:23.181123444 +0000
@@ -5,6 +5,7 @@
  *
  *  Copyright (C) 1999, 2000 Red Hat, Inc.
  *  Copyright (C) 2000, 2001 Eazel, Inc.
+ *  Copyright (c) 2020, Dynamic Systems, Inc.
  *
  *  Caja is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public License as
diff -ur caja-1.22.1/src/caja-desktop-window.c caja-1.22.1/src/caja-desktop-window.c
--- caja-1.22.1/src/caja-desktop-window.c	2020-04-24 23:05:53.002344387 +0000
+++ caja-1.22.1/src/caja-desktop-window.c	2020-05-26 16:30:31.882851374 +0000
@@ -4,6 +4,7 @@
  * Caja
  *
  * Copyright (C) 2000 Eazel, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Caja is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur caja-1.22.1/src/caja-main.c caja-1.22.1/src/caja-main.c
--- caja-1.22.1/src/caja-main.c	2020-04-24 23:05:53.003100990 +0000
+++ caja-1.22.1/src/caja-main.c	2020-05-26 16:30:40.644474910 +0000
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 1999, 2000 Red Hat, Inc.
  * Copyright (C) 1999, 2000 Eazel, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Caja is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur caja-1.22.1/src/caja-window-manage-views.c caja-1.22.1/src/caja-window-manage-views.c
--- caja-1.22.1/src/caja-window-manage-views.c	2020-04-24 23:05:53.004078421 +0000
+++ caja-1.22.1/src/caja-window-manage-views.c	2020-05-26 16:30:56.597029448 +0000
@@ -5,6 +5,7 @@
  *
  *  Copyright (C) 1999, 2000 Red Hat, Inc.
  *  Copyright (C) 1999, 2000, 2001 Eazel, Inc.
+ *  Copyright (c) 2020, Dynamic Systems, Inc.
  *
  *  Caja is free software; you can redistribute it and/or
  *  modify it under the terms of the GNU General Public
diff -ur caja-1.22.1/src/caja-zfs-bar.c caja-1.22.1/src/caja-zfs-bar.c
--- caja-1.22.1/src/caja-zfs-bar.c	2020-04-24 23:05:53.004869490 +0000
+++ caja-1.22.1/src/caja-zfs-bar.c	2020-05-26 16:31:07.083383780 +0000
@@ -1,5 +1,6 @@
 /* 
  * Copyright (c) 2008, 2011, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  */
 
diff -ur caja-1.22.1/src/file-manager/fm-directory-view.c caja-1.22.1/src/file-manager/fm-directory-view.c
--- caja-1.22.1/src/file-manager/fm-directory-view.c	2020-04-24 23:05:53.007556266 +0000
+++ caja-1.22.1/src/file-manager/fm-directory-view.c	2020-05-26 16:31:20.951763967 +0000
@@ -4,6 +4,7 @@
  *
  * Copyright (C) 1999, 2000  Free Software Foundation
  * Copyright (C) 2000, 2001  Eazel, Inc.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur caja-1.22.1/src/file-manager/fm-properties-window.c caja-1.22.1/src/file-manager/fm-properties-window.c
--- caja-1.22.1/src/file-manager/fm-properties-window.c	2020-04-24 23:05:53.008866567 +0000
+++ caja-1.22.1/src/file-manager/fm-properties-window.c	2020-05-26 16:31:34.646335191 +0000
@@ -3,6 +3,7 @@
 /* fm-properties-window.c - window that lets user modify file properties
 
    Copyright (C) 2000 Eazel, Inc.
+   Copyright (c) 2020, Dynamic Systems, Inc.
 
    The Mate Library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public License as
