diff -ur gnome-power-manager-2.24.4/Makefile.in gnome-power-manager-2.24.4/Makefile.in
--- gnome-power-manager-2.24.4/Makefile.in	2020-05-01 15:52:51.775884140 +0000
+++ gnome-power-manager-2.24.4/Makefile.in	2020-05-26 16:45:30.272689481 +0000
@@ -2,6 +2,7 @@
 # @configure_input@
 
 # Copyright (C) 1994-2018 Free Software Foundation, Inc.
+# Copyright (c) 2020, Dynamic Systems, Inc.
 
 # This Makefile.in is free software; the Free Software Foundation
 # gives unlimited permission to copy and/or distribute it,
diff -ur gnome-power-manager-2.24.4/configure gnome-power-manager-2.24.4/configure
--- gnome-power-manager-2.24.4/configure	2020-05-01 15:52:49.562118104 +0000
+++ gnome-power-manager-2.24.4/configure	2020-05-26 16:46:07.245175229 +0000
@@ -4,6 +4,7 @@
 #
 #
 # Copyright (C) 1992-1996, 1998-2012 Free Software Foundation, Inc.
+# Copyright (c) 2020, Dynamic Systems, Inc.
 #
 #
 # This configure script is free software; the Free Software Foundation
diff -ur gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.c gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.c
--- gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.c	2020-05-01 15:52:40.746632650 +0000
+++ gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.c	2020-05-26 16:46:20.342130123 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.h gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.h
--- gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.h	2020-05-01 15:52:40.747517842 +0000
+++ gnome-power-manager-2.24.4/libhal-glib/libhal-gcpufreq.h	2020-05-26 16:46:25.869876944 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/libidletime/libidletime.c gnome-power-manager-2.24.4/libidletime/libidletime.c
--- gnome-power-manager-2.24.4/libidletime/libidletime.c	2020-05-01 15:52:40.749052862 +0000
+++ gnome-power-manager-2.24.4/libidletime/libidletime.c	2020-05-26 16:46:32.032545157 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/Makefile.in gnome-power-manager-2.24.4/src/Makefile.in
--- gnome-power-manager-2.24.4/src/Makefile.in	2020-05-01 15:52:53.816950496 +0000
+++ gnome-power-manager-2.24.4/src/Makefile.in	2020-05-26 16:46:40.777038855 +0000
@@ -2,6 +2,7 @@
 # @configure_input@
 
 # Copyright (C) 1994-2018 Free Software Foundation, Inc.
+# Copyright (c) 2020, Dynamic Systems, Inc.
 
 # This Makefile.in is free software; the Free Software Foundation
 # gives unlimited permission to copy and/or distribute it,
diff -ur gnome-power-manager-2.24.4/src/egg-array-float.c gnome-power-manager-2.24.4/src/egg-array-float.c
--- gnome-power-manager-2.24.4/src/egg-array-float.c	2020-05-01 15:52:40.757353684 +0000
+++ gnome-power-manager-2.24.4/src/egg-array-float.c	2020-05-26 16:46:54.676343167 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007-2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/egg-array-float.h gnome-power-manager-2.24.4/src/egg-array-float.h
--- gnome-power-manager-2.24.4/src/egg-array-float.h	2020-05-01 15:52:40.758121169 +0000
+++ gnome-power-manager-2.24.4/src/egg-array-float.h	2020-05-26 16:47:01.214047406 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007-2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/egg-debug.h gnome-power-manager-2.24.4/src/egg-debug.h
--- gnome-power-manager-2.24.4/src/egg-debug.h	2020-05-01 15:52:40.759072851 +0000
+++ gnome-power-manager-2.24.4/src/egg-debug.h	2020-05-26 16:47:07.717062919 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007-2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/egg-precision.c gnome-power-manager-2.24.4/src/egg-precision.c
--- gnome-power-manager-2.24.4/src/egg-precision.c	2020-05-01 15:52:40.759863654 +0000
+++ gnome-power-manager-2.24.4/src/egg-precision.c	2020-05-26 16:47:43.188766196 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007-2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/egg-precision.h gnome-power-manager-2.24.4/src/egg-precision.h
--- gnome-power-manager-2.24.4/src/egg-precision.h	2020-05-01 15:52:40.760606606 +0000
+++ gnome-power-manager-2.24.4/src/egg-precision.h	2020-05-26 16:47:54.895977663 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007-2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-array-float.c gnome-power-manager-2.24.4/src/gpm-array-float.c
--- gnome-power-manager-2.24.4/src/gpm-array-float.c	2020-05-01 15:52:40.761550139 +0000
+++ gnome-power-manager-2.24.4/src/gpm-array-float.c	2020-05-26 16:47:59.852563022 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-array-float.h gnome-power-manager-2.24.4/src/gpm-array-float.h
--- gnome-power-manager-2.24.4/src/gpm-array-float.h	2020-05-01 15:52:40.762389758 +0000
+++ gnome-power-manager-2.24.4/src/gpm-array-float.h	2020-05-26 16:48:05.573361372 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-brightness-xrandr.c gnome-power-manager-2.24.4/src/gpm-brightness-xrandr.c
--- gnome-power-manager-2.24.4/src/gpm-brightness-xrandr.c	2020-05-01 15:52:40.763440055 +0000
+++ gnome-power-manager-2.24.4/src/gpm-brightness-xrandr.c	2020-05-26 16:48:14.012234348 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-button.c gnome-power-manager-2.24.4/src/gpm-button.c
--- gnome-power-manager-2.24.4/src/gpm-button.c	2020-05-01 15:52:40.764451956 +0000
+++ gnome-power-manager-2.24.4/src/gpm-button.c	2020-05-26 16:48:20.036980162 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-cell-unit.c gnome-power-manager-2.24.4/src/gpm-cell-unit.c
--- gnome-power-manager-2.24.4/src/gpm-cell-unit.c	2020-05-01 15:52:40.765358796 +0000
+++ gnome-power-manager-2.24.4/src/gpm-cell-unit.c	2020-05-26 16:48:26.337374210 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-common.c gnome-power-manager-2.24.4/src/gpm-common.c
--- gnome-power-manager-2.24.4/src/gpm-common.c	2020-05-01 15:52:40.766248123 +0000
+++ gnome-power-manager-2.24.4/src/gpm-common.c	2020-05-26 16:48:41.795409578 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-conf.h gnome-power-manager-2.24.4/src/gpm-conf.h
--- gnome-power-manager-2.24.4/src/gpm-conf.h	2020-05-01 15:52:40.767175541 +0000
+++ gnome-power-manager-2.24.4/src/gpm-conf.h	2020-05-26 16:48:50.316027022 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-control.c gnome-power-manager-2.24.4/src/gpm-control.c
--- gnome-power-manager-2.24.4/src/gpm-control.c	2020-05-01 15:52:40.768771633 +0000
+++ gnome-power-manager-2.24.4/src/gpm-control.c	2020-05-26 16:48:57.115613658 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-control.h gnome-power-manager-2.24.4/src/gpm-control.h
--- gnome-power-manager-2.24.4/src/gpm-control.h	2020-05-01 15:52:40.769609614 +0000
+++ gnome-power-manager-2.24.4/src/gpm-control.h	2020-05-26 16:49:03.609568345 +0000
@@ -2,6 +2,7 @@
  *
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-cpufreq.c gnome-power-manager-2.24.4/src/gpm-cpufreq.c
--- gnome-power-manager-2.24.4/src/gpm-cpufreq.c	2020-05-01 15:52:40.770448380 +0000
+++ gnome-power-manager-2.24.4/src/gpm-cpufreq.c	2020-05-26 16:49:10.906581954 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-cpufreq.h gnome-power-manager-2.24.4/src/gpm-cpufreq.h
--- gnome-power-manager-2.24.4/src/gpm-cpufreq.h	2020-05-01 15:52:40.771231938 +0000
+++ gnome-power-manager-2.24.4/src/gpm-cpufreq.h	2020-05-26 16:49:16.176803701 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-dpms.c gnome-power-manager-2.24.4/src/gpm-dpms.c
--- gnome-power-manager-2.24.4/src/gpm-dpms.c	2020-05-01 15:52:40.772538089 +0000
+++ gnome-power-manager-2.24.4/src/gpm-dpms.c	2020-05-26 16:49:21.826388766 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-feedback-widget.c gnome-power-manager-2.24.4/src/gpm-feedback-widget.c
--- gnome-power-manager-2.24.4/src/gpm-feedback-widget.c	2020-05-01 15:52:40.773817728 +0000
+++ gnome-power-manager-2.24.4/src/gpm-feedback-widget.c	2020-05-26 16:49:29.258141152 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-graph-widget.c gnome-power-manager-2.24.4/src/gpm-graph-widget.c
--- gnome-power-manager-2.24.4/src/gpm-graph-widget.c	2020-05-01 15:52:40.777223058 +0000
+++ gnome-power-manager-2.24.4/src/gpm-graph-widget.c	2020-05-26 16:49:35.680537352 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-graph-widget.h gnome-power-manager-2.24.4/src/gpm-graph-widget.h
--- gnome-power-manager-2.24.4/src/gpm-graph-widget.h	2020-05-01 15:52:40.778256605 +0000
+++ gnome-power-manager-2.24.4/src/gpm-graph-widget.h	2020-05-26 16:49:41.900464334 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-main.c gnome-power-manager-2.24.4/src/gpm-main.c
--- gnome-power-manager-2.24.4/src/gpm-main.c	2020-05-01 15:52:40.779316587 +0000
+++ gnome-power-manager-2.24.4/src/gpm-main.c	2020-05-26 16:49:47.818140077 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Taken in part from:
  *  - lshal   (C) 2003 David Zeuthen, <david@fubar.dk>
diff -ur gnome-power-manager-2.24.4/src/gpm-manager.c gnome-power-manager-2.24.4/src/gpm-manager.c
--- gnome-power-manager-2.24.4/src/gpm-manager.c	2020-05-01 15:52:40.780477579 +0000
+++ gnome-power-manager-2.24.4/src/gpm-manager.c	2020-05-26 16:49:52.689414853 +0000
@@ -2,6 +2,7 @@
  *
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-notify.c gnome-power-manager-2.24.4/src/gpm-notify.c
--- gnome-power-manager-2.24.4/src/gpm-notify.c	2020-05-01 15:52:40.781530382 +0000
+++ gnome-power-manager-2.24.4/src/gpm-notify.c	2020-05-26 16:50:08.481724260 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-notify.h gnome-power-manager-2.24.4/src/gpm-notify.h
--- gnome-power-manager-2.24.4/src/gpm-notify.h	2020-05-01 15:52:40.782356007 +0000
+++ gnome-power-manager-2.24.4/src/gpm-notify.h	2020-05-26 16:50:15.375604971 +0000
@@ -2,6 +2,7 @@
  *
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-point-obj.c gnome-power-manager-2.24.4/src/gpm-point-obj.c
--- gnome-power-manager-2.24.4/src/gpm-point-obj.c	2020-05-01 15:52:40.783134034 +0000
+++ gnome-power-manager-2.24.4/src/gpm-point-obj.c	2020-05-26 16:50:21.297965990 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
diff -ur gnome-power-manager-2.24.4/src/gpm-point-obj.h gnome-power-manager-2.24.4/src/gpm-point-obj.h
--- gnome-power-manager-2.24.4/src/gpm-point-obj.h	2020-05-01 15:52:40.783969515 +0000
+++ gnome-power-manager-2.24.4/src/gpm-point-obj.h	2020-05-26 16:50:27.686831035 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
diff -ur gnome-power-manager-2.24.4/src/gpm-prefs-core.c gnome-power-manager-2.24.4/src/gpm-prefs-core.c
--- gnome-power-manager-2.24.4/src/gpm-prefs-core.c	2020-05-01 15:52:40.786770538 +0000
+++ gnome-power-manager-2.24.4/src/gpm-prefs-core.c	2020-05-26 16:50:35.338723045 +0000
@@ -3,6 +3,7 @@
  * Copyright (C) 2005 Jaap Haitsma <jaap@haitsma.org>
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-profile.h gnome-power-manager-2.24.4/src/gpm-profile.h
--- gnome-power-manager-2.24.4/src/gpm-profile.h	2020-05-01 15:52:40.787607412 +0000
+++ gnome-power-manager-2.24.4/src/gpm-profile.h	2020-05-26 16:50:45.256231291 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-screensaver.c gnome-power-manager-2.24.4/src/gpm-screensaver.c
--- gnome-power-manager-2.24.4/src/gpm-screensaver.c	2020-05-01 15:52:40.788656738 +0000
+++ gnome-power-manager-2.24.4/src/gpm-screensaver.c	2020-05-26 16:50:51.736320436 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-session.c gnome-power-manager-2.24.4/src/gpm-session.c
--- gnome-power-manager-2.24.4/src/gpm-session.c	2020-05-01 15:52:40.789510503 +0000
+++ gnome-power-manager-2.24.4/src/gpm-session.c	2020-05-26 16:50:56.591280601 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2008-2009 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-session.h gnome-power-manager-2.24.4/src/gpm-session.h
--- gnome-power-manager-2.24.4/src/gpm-session.h	2020-05-01 15:52:40.790270459 +0000
+++ gnome-power-manager-2.24.4/src/gpm-session.h	2020-05-26 16:51:02.543022226 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2008 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-statistics-core.c gnome-power-manager-2.24.4/src/gpm-statistics-core.c
--- gnome-power-manager-2.24.4/src/gpm-statistics-core.c	2020-05-01 15:52:40.792028492 +0000
+++ gnome-power-manager-2.24.4/src/gpm-statistics-core.c	2020-05-26 16:51:08.714051572 +0000
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2006-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
diff -ur gnome-power-manager-2.24.4/src/gpm-tray-icon.c gnome-power-manager-2.24.4/src/gpm-tray-icon.c
--- gnome-power-manager-2.24.4/src/gpm-tray-icon.c	2020-05-01 15:52:40.793196126 +0000
+++ gnome-power-manager-2.24.4/src/gpm-tray-icon.c	2020-05-26 16:51:20.688533885 +0000
@@ -2,6 +2,7 @@
  *
  * Copyright (C) 2005 William Jon McCann <mccann@jhu.edu>
  * Copyright (C) 2005-2007 Richard Hughes <richard@hughsie.com>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Licensed under the GNU General Public License Version 2
  *
