diff -ur xtrans-1.3.5/Xtranssock.c xtrans-1.3.5/Xtranssock.c
--- xtrans-1.3.5/Xtranssock.c	2020-05-01 16:41:30.439750301 +0000
+++ xtrans-1.3.5/Xtranssock.c	2020-05-26 16:54:18.379779344 +0000
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2002, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * Permission is hereby granted, free of charge, to any person obtaining a
  * copy of this software and associated documentation files (the "Software"),
