diff -ur gnome-screenshot-3.22.0/src/screenshot-utils.c gnome-screenshot-3.22.0/src/screenshot-utils.c
--- gnome-screenshot-3.22.0/src/screenshot-utils.c	2020-05-01 15:38:39.556932837 +0000
+++ gnome-screenshot-3.22.0/src/screenshot-utils.c	2020-05-26 16:51:36.167372086 +0000
@@ -2,6 +2,7 @@
  *
  * Copyright (C) 2001-2006  Jonathan Blandford <jrb@alum.mit.edu>
  * Copyright (C) 2008 Cosimo Cecchi <cosimoc@gnome.org>
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public
