diff -ur marco-1.22.1/src/core/atomnames.h marco-1.22.1/src/core/atomnames.h
--- marco-1.22.1/src/core/atomnames.h	2020-05-21 19:52:45.417997716 +0000
+++ marco-1.22.1/src/core/atomnames.h	2020-05-26 16:32:38.102683557 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2003, 2004 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
  * Copyright (C) 2008 Thomas Thurman
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/display-private.h marco-1.22.1/src/core/display-private.h
--- marco-1.22.1/src/core/display-private.h	2020-05-21 19:52:45.418804330 +0000
+++ marco-1.22.1/src/core/display-private.h	2020-05-26 16:32:45.464067533 +0000
@@ -5,6 +5,7 @@
  * Copyright (C) 2002 Red Hat, Inc.
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/display.c marco-1.22.1/src/core/display.c
--- marco-1.22.1/src/core/display.c	2020-05-21 19:52:45.421129679 +0000
+++ marco-1.22.1/src/core/display.c	2020-05-26 16:32:52.241487786 +0000
@@ -5,6 +5,7 @@
  * Copyright (C) 2002, 2003, 2004 Red Hat, Inc.
  * Copyright (C) 2003, 2004 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/frame.c marco-1.22.1/src/core/frame.c
--- marco-1.22.1/src/core/frame.c	2020-05-21 19:52:45.421845849 +0000
+++ marco-1.22.1/src/core/frame.c	2020-05-26 16:33:00.847006808 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2003, 2004 Red Hat, Inc.
  * Copyright (C) 2005 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/keybindings.c marco-1.22.1/src/core/keybindings.c
--- marco-1.22.1/src/core/keybindings.c	2020-05-21 19:52:45.423188000 +0000
+++ marco-1.22.1/src/core/keybindings.c	2020-05-26 16:33:08.071000609 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2002 Red Hat Inc.
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/main.c marco-1.22.1/src/core/main.c
--- marco-1.22.1/src/core/main.c	2020-05-21 19:52:45.423836286 +0000
+++ marco-1.22.1/src/core/main.c	2020-05-26 16:33:14.742711371 +0000
@@ -5,6 +5,7 @@
 /*
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/prefs.c marco-1.22.1/src/core/prefs.c
--- marco-1.22.1/src/core/prefs.c	2020-05-21 19:52:45.426133361 +0000
+++ marco-1.22.1/src/core/prefs.c	2020-05-26 16:33:21.199535479 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2001 Havoc Pennington, Copyright (C) 2002 Red Hat Inc.
  * Copyright (C) 2006 Elijah Newren
  * Copyright (C) 2008 Thomas Thurman
+ * Copyright (c) 2021, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/screen-private.h marco-1.22.1/src/core/screen-private.h
--- marco-1.22.1/src/core/screen-private.h	2020-05-21 19:52:45.426717746 +0000
+++ marco-1.22.1/src/core/screen-private.h	2020-05-26 16:33:31.405662221 +0000
@@ -13,6 +13,7 @@
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/screen.c marco-1.22.1/src/core/screen.c
--- marco-1.22.1/src/core/screen.c	2020-05-21 19:52:45.427846145 +0000
+++ marco-1.22.1/src/core/screen.c	2020-05-26 16:33:43.743512172 +0000
@@ -9,6 +9,7 @@
  * Copyright (C) 2001 Dominik Vogt, Matthias Clasen, and fvwm2 team
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/trusted-pics.h marco-1.22.1/src/core/trusted-pics.h
--- marco-1.22.1/src/core/trusted-pics.h	2020-05-21 19:52:45.428616713 +0000
+++ marco-1.22.1/src/core/trusted-pics.h	2020-05-26 16:33:57.905980632 +0000
@@ -2,6 +2,7 @@
 
 /* 
  * Copyright (C) 2005 Erwann Chenede
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  */
 
 #ifndef TRUSTED_PICS_H
diff -ur marco-1.22.1/src/core/trusted.c marco-1.22.1/src/core/trusted.c
--- marco-1.22.1/src/core/trusted.c	2020-05-21 19:52:45.429712972 +0000
+++ marco-1.22.1/src/core/trusted.c	2020-05-26 16:34:08.398873555 +0000
@@ -2,6 +2,7 @@
 
 /* 
  * Copyright (C) 2005 Erwann Chenede
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  */
 
 #include <config.h>
diff -ur marco-1.22.1/src/core/trusted.h marco-1.22.1/src/core/trusted.h
--- marco-1.22.1/src/core/trusted.h	2020-05-21 19:52:45.430506527 +0000
+++ marco-1.22.1/src/core/trusted.h	2020-05-26 16:34:14.815383616 +0000
@@ -2,6 +2,7 @@
 
 /* 
  * Copyright (C) 2005 Erwann Chenede
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  */
 
 #ifndef TRUSTED_H
diff -ur marco-1.22.1/src/core/window-props.c marco-1.22.1/src/core/window-props.c
--- marco-1.22.1/src/core/window-props.c	2020-05-21 19:52:45.431548004 +0000
+++ marco-1.22.1/src/core/window-props.c	2020-05-26 16:34:24.238050034 +0000
@@ -18,6 +18,7 @@
  * Copyright (C) 2001, 2002, 2003 Red Hat, Inc.
  * Copyright (C) 2004, 2005 Elijah Newren
  * Copyright (C) 2009 Thomas Thurman
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/window.c marco-1.22.1/src/core/window.c
--- marco-1.22.1/src/core/window.c	2020-05-21 19:52:45.434013790 +0000
+++ marco-1.22.1/src/core/window.c	2020-05-26 16:34:32.686810325 +0000
@@ -7,6 +7,7 @@
  * Copyright (C) 2002, 2003 Red Hat, Inc.
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004-2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/workspace.c marco-1.22.1/src/core/workspace.c
--- marco-1.22.1/src/core/workspace.c	2020-05-21 19:52:45.434615199 +0000
+++ marco-1.22.1/src/core/workspace.c	2020-05-26 16:34:40.066967269 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2003 Rob Adams
  * Copyright (C) 2004, 2005 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/core/workspace.h marco-1.22.1/src/core/workspace.h
--- marco-1.22.1/src/core/workspace.h	2020-05-21 19:52:45.435090476 +0000
+++ marco-1.22.1/src/core/workspace.h	2020-05-26 16:34:49.565833188 +0000
@@ -13,6 +13,7 @@
 /*
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2004, 2005 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/include/prefs.h marco-1.22.1/src/include/prefs.h
--- marco-1.22.1/src/include/prefs.h	2020-05-21 19:52:45.435707673 +0000
+++ marco-1.22.1/src/include/prefs.h	2020-05-26 16:34:56.932505941 +0000
@@ -5,6 +5,7 @@
 /*
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/include/ui.h marco-1.22.1/src/include/ui.h
--- marco-1.22.1/src/include/ui.h	2020-05-21 19:52:45.436185687 +0000
+++ marco-1.22.1/src/include/ui.h	2020-05-26 16:35:03.647701672 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2001 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/tools/marco-window-demo.c marco-1.22.1/src/tools/marco-window-demo.c
--- marco-1.22.1/src/tools/marco-window-demo.c	2020-05-21 19:52:45.436844159 +0000
+++ marco-1.22.1/src/tools/marco-window-demo.c	2020-05-26 16:35:13.573756514 +0000
@@ -2,6 +2,7 @@
 
 /*
  * Copyright (C) 2002 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/frames.c marco-1.22.1/src/ui/frames.c
--- marco-1.22.1/src/ui/frames.c	2020-05-21 19:52:45.437790888 +0000
+++ marco-1.22.1/src/ui/frames.c	2020-05-26 16:35:21.383487137 +0000
@@ -6,6 +6,7 @@
  * Copyright (C) 2001 Havoc Pennington
  * Copyright (C) 2003 Red Hat, Inc.
  * Copyright (C) 2005, 2006 Elijah Newren
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/frames.h marco-1.22.1/src/ui/frames.h
--- marco-1.22.1/src/ui/frames.h	2020-05-21 19:52:45.438397006 +0000
+++ marco-1.22.1/src/ui/frames.h	2020-05-26 16:35:26.859118124 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2001 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/preview-widget.c marco-1.22.1/src/ui/preview-widget.c
--- marco-1.22.1/src/ui/preview-widget.c	2020-05-21 19:53:10.851942127 +0000
+++ marco-1.22.1/src/ui/preview-widget.c	2020-05-26 16:35:34.211551482 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2002 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/theme-viewer.c marco-1.22.1/src/ui/theme-viewer.c
--- marco-1.22.1/src/ui/theme-viewer.c	2020-05-21 19:52:45.440592802 +0000
+++ marco-1.22.1/src/ui/theme-viewer.c	2020-05-26 16:35:57.548034093 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2002 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/theme.c marco-1.22.1/src/ui/theme.c
--- marco-1.22.1/src/ui/theme.c	2020-05-21 19:52:45.443479422 +0000
+++ marco-1.22.1/src/ui/theme.c	2020-05-26 16:36:04.044056927 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2001 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/theme.h marco-1.22.1/src/ui/theme.h
--- marco-1.22.1/src/ui/theme.h	2020-05-21 19:52:45.444667392 +0000
+++ marco-1.22.1/src/ui/theme.h	2020-05-26 16:36:10.739909390 +0000
@@ -4,6 +4,7 @@
 
 /*
  * Copyright (C) 2001 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff -ur marco-1.22.1/src/ui/ui.c marco-1.22.1/src/ui/ui.c
--- marco-1.22.1/src/ui/ui.c	2020-05-21 19:52:45.445225555 +0000
+++ marco-1.22.1/src/ui/ui.c	2020-05-26 16:36:29.121862092 +0000
@@ -2,6 +2,7 @@
 
 /*
  * Copyright (C) 2002 Havoc Pennington
+ * Copyright (c) 2020, Dynamic Systems, Inc.
  * stock icon code Copyright (C) 2002 Jorn Baayen <jorn@nl.linux.org>
  *
  * This program is free software; you can redistribute it and/or
