diff -ur data/mate-power-preferences.desktop.in data2/mate-power-preferences.desktop.in
--- a/data/mate-power-preferences.desktop.in	2019-04-22 21:43:58.000000000 +0000
+++ b/data/mate-power-preferences.desktop.in	2020-07-11 17:11:58.101792288 +0000
@@ -2,7 +2,7 @@
 _Name=Power Management
 _Comment=Configure power management
 Icon=mate-power-manager
-Exec=mate-power-preferences
+Exec=mate-power-manager
 Terminal=false
 Type=Application
 Categories=Settings;HardwareSettings;
diff -ur data/mate-power-preferences.desktop.in.in data2/mate-power-preferences.desktop.in.in
--- a/data/mate-power-preferences.desktop.in.in	2019-04-22 21:40:32.000000000 +0000
+++ b/data/mate-power-preferences.desktop.in.in	2020-07-11 17:11:58.102137060 +0000
@@ -2,7 +2,7 @@
 _Name=Power Management
 _Comment=Configure power management
 Icon=mate-power-manager
-Exec=mate-power-preferences
+Exec=mate-power-manager
 Terminal=false
 Type=Application
 Categories=Settings;HardwareSettings;
